wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -88,22 +94,61 @@ public void apply(
             || OPENSHIFT_TOOL_TYPE.equals(k8sTool.getType()),
         format("Plugin must have `%s` or `%s` type", KUBERNETES_TOOL_TYPE, OPENSHIFT_TOOL_TYPE));
 
-    String recipeFileContent = retrieveContent(k8sTool, contentProvider);
-
-    List<HasMetadata> list = unmarshal(k8sTool, recipeFileContent);
+    String toolContent = retrieveContent(k8sTool, contentProvider);
 
+    List<HasMetadata> toolObjects = unmarshalToolObjects(k8sTool, toolContent);
     if (!k8sTool.getSelector().isEmpty()) {
-      list = filter(list, k8sTool.getSelector());
+      toolObjects = filter(toolObjects, k8sTool.getSelector());
     }
 
-    estimateCommandsMachineName(workspaceConfig, k8sTool, list);
+    K8sToolEnvironmentBuilder environmentBuilder =
+        envBuilderFactory.create(k8sTool, workspaceConfig);
+
+    List<PodData> toolPods = removePods(toolObjects);
```

Comment:
I don't like how the logic with multiple deployment/pod removal is spread over this method, `PodsMerger` and `K8sToolEnvironmentBuilder` in a quite intertwined way.

The logic should be simple (and documented ;)

1. extract pods from the source
1. remove all deployments from the source
1. create a new deployment
1. merge all pods into 1
1. add the new pod to the new deployment
1. sacrifice a lamb
1. pray it all works

Reply:
It trying to move Pods merging logic on infra level. It should look much simpler here.

Reply:
is not actual anymore. Please take a look how it is done in `KubernetesEnvironmentFactory`